lessons/advanced/gen-stage.md

```diff
@@ -16,46 +16,353 @@ So what is GenStage?  From the official documentation, it is a "specification an
 
-To better understand how this works, let's visualize a simple producer-consumer flow:
+We could go into the technical and theoretical implications of this, but instead lets try a pragmatic approach to really just get it to work.
 
+First, Let's imagine we have a server that constantly emits numbers.
+It starts at the state of the number we give it, then counts up in one from htere onward.
+This is what we would call our producer.
+Each time it emits a number, this is an event, and we want to handle it with a consumer.
+A consumer simply takes what a producer emits and does something to it.
+In our case, we will display the count.
+There is a lot more to GenStage at a technical and applied level, but we will build up on the specifics and definitions further in later lessons, for now we just want a running example we can build up on.
+
+## Getting Started: A Sample GenStage Project
+We'll begin by generating a simple project that has a supervision tree:
+
+```shell
+$ mix new genstage_example --sup
+$ cd genstage_example
+```
+
+Let's set up some basic things for the future of our application.
+Since GenStage is generally used as a transformation pipeline, lets imagine we have a background worker of some sort.
+This worker will need to persist whatever it changes, so we should get a database set up, but we can worry about that in a later lesson.
+To start, all we need to do is add `gen_stage` to our deps.
+
+```elixir
+. . .
+  defp deps do
+    [
+      {:gen_stage, "~> 0.7},
+    ]
+  end
+. . .
+```
+
+Now, we should fetch our dependencies and compile before we start setup:
+
+```shell
+$ mix do deps.get, compile
+```
+
+Lets build a producer, our simple beginning building block to help us utilize this new tool!
+
+## Building A Producer
+To get started what we want to do is create a producer that emits a constant stream of events for our consumer to handle.
+This is quite simple with the rudimentary example of a counter.
+Let's create a namespaced directory under `lib` and then go from there, this way our module naming matches our names of the modules themselves:
+
+```shell
+$ mkdir lib/genstage_example
+$ touch lib/genstage_example/producer.ex
+```
+
+Now we can add the code:
+
+```elixir
+defmodule GenstageExample.Producer do
+  alias Experimental.GenStage
+  use GenStage
+
+  def start_link do
+    GenStage.start_link(__MODULE__, 0, name: __MODULE__)
+                                       # naming allows us to handle failure
+  end
+
+  def init(counter) do
+    {:producer, counter}
+  end
+
+  def handle_demand(demand, state) do
+    events = Enum.to_list(state..state + demand - 1)
+    {:noreply, events, (state + demand)}
+  end
+end
+```
+[working commit](https://github.com/ybur-yug/genstage_example/commit/57418160a1e5cbb408e32184d913a12e207166f8)
```

Comment:
I know we havent done this before but I think (especially for advanced topics) that it could help a lot for those who might end up stuck.